crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -0,0 +1,863 @@
+use super::{events::meta_to_event_index, Metrics, Postgres};
+use anyhow::{anyhow, bail, Context, Result};
+use contracts::cowswap_onchain_orders::{
+    event_data::OrderPlacement as ContractOrderPlacement, Event as ContractEvent,
+};
+use database::{
+    byte_array::ByteArray, events::EventIndex, onchain_broadcasted_orders::OnchainOrderPlacement,
+    orders::Order, PgTransaction,
+};
+use ethcontract::{Event as EthContractEvent, H160};
+use futures::{stream, StreamExt};
+use itertools::multiunzip;
+use model::{
+    order::{
+        buy_token_destination_from_contract_bytes, kind_from_contract_constant,
+        sell_token_source_from_contract_bytes, OrderUid,
+    },
+    signature::SigningScheme,
+    DomainSeparator,
+    {app_id::AppId, order::OrderData},
+};
+use number_conversions::u256_to_big_decimal;
+use shared::{
+    db_order_conversions::{
+        buy_token_destination_into, order_kind_into, sell_token_source_into, signing_scheme_into,
+    },
+    event_handling::EventStoring,
+    order_quoting::{OrderQuoting, Quote, QuoteSearchParameters},
+    order_validation::{
+        convert_signing_scheme_into_quote_kind, get_quote_and_check_fee,
+        is_order_outside_market_price,
+    },
+};
+use std::collections::HashMap;
+
+pub struct OnchainOrderParser<T: Send + Sync, W: Send + Sync> {
+    db: Postgres,
+    quoter: Box<dyn OrderQuoting>,
+    custom_onchain_data_parser: Box<dyn OnchainOrderParsing<T, W>>,
+    domain_separator: DomainSeparator,
+    settlement_contract: H160,
+}
+
+impl<T: Send + Sync, W: Send + Sync> OnchainOrderParser<T, W> {
+    pub fn new(
+        db: Postgres,
+        quoter: Box<dyn OrderQuoting>,
+        custom_onchain_data_parser: Box<dyn OnchainOrderParsing<T, W>>,
+        domain_separator: DomainSeparator,
+        settlement_contract: H160,
+    ) -> Self {
+        OnchainOrderParser {
+            db,
+            quoter,
+            custom_onchain_data_parser,
+            domain_separator,
+            settlement_contract,
+        }
+    }
+}
+
+// The following struct describes the return type from the custom order parsing logic.
+// All parser must return a quote_id, as this is currently required by the protocol.
+pub struct OnchainOrderCustomData<T> {
+    quote_id: i64,
+    additional_data: Option<T>,
+}
+
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+// The following trait allows to implement custom onchain order parsing for differently placed
+// orders. E.g. there will be a implementation for ethflow and presign orders.
+// For each of the customs types, the trait allows to implement parsing the on-chain data and
+// storing the event data
+
+// The generic EventData stores the result of the custom event parsing
+// The generic EvenDataForDB contains the prepared data that will be appended to the database
+pub trait OnchainOrderParsing<EventData: Send + Sync + Clone, EventDataForDB: Send + Sync>:
+    Send + Sync
```

Comment:
Just a suggestion, but fine as is:
```suggestion
pub trait OnchainOrderParsing<EventData, EventDataForDB>: Send + Sync
where
    EventData: Send + Sync + Clone,
    EventDataForDB: Send + Sync,
```

Also, as a naming nit maybe `EventRow` instead of `EventDataForDB`?